### PR TITLE
POC: container live patching

### PR DESCRIPTION
**Pull request by brauner**
This adds set_running_config_item() which is the analogue of
get_running_config_item(). In essence it allows a caller to livepatch the
container's in-memory configuration. This POC is severly limited. Here are the
most obvious ones:
- Only the container's in-memory config can be updated but no further actions
  (e.g. on-disk actions) are made.
- Only keys in the "lxc.net." namespace can be changed. This POC also allows
  updating an existing network. For example it allows to change the network
  type of an existing network. This is obviously nonsense and in a non-POC
  implementation this should be blocked.

Use Case:
Callers can hotplug a new network for the container. For example, LXD can
create a pair of veth devices in the host and in the container and add it to
the container's in-memory config. This means, the container can later be
queried for the name of the device later on etc. Note that liblxc will
currently not delete hotplugged network devices on container shutdown since it
won't have the ifindex of the container.

Relates to lxc/lxd#3920 .

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

File: src/lxc/lxccontainer.c [hunk 1]
```diff
@@ -2765,29 +2765,47 @@ static bool do_lxcapi_destroy_with_snapshots(struct lxc_container *c)
...
+		do_clear_unexp_config_line(conf, key);
+	else
+		bret = do_append_unexp_config_line(conf, key, v);
+	if (!bret)
+		return -ENOMEM;
+
+	return 0;
+}
+
+static bool do_set_config_item_locked(struct lxc_container *c, const char *key,
```
Comment (hallyn) [inline src/lxc/lxccontainer.c]: Generally fn() would call do_fn(), right?  This do_fn() and lxc_fn() is confusing and might become problematic...  Unfortunately I don't have any better ideas for naming.  Do you?  (If not, then I guess we leave it as is)

File: src/tests/livepatch.c [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	int ret = EXIT_FAILURE;
+
+	c = lxc_container_new("livepatch", NULL);
+	if (!c) {
+		lxc_error("%s", "Failed to create container \"livepatch\"");
+		exit(ret);
+	}
+
+	if (c->is_defined(c)) {
+		lxc_error("%s\n", "Container \"livepatch\" is not defined");
```
Comment (hallyn) [inline src/tests/livepatch.c]: is *not* defined?

File: src/tests/livepatch.c [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+		free(value);
+		goto on_error_destroy;
+	}
+	free(value);
+
+	/* Remove in-memory value. */
+	if (!c->set_running_config_item(c, "lxc.net.1.type", "")) {
+		lxc_error("%s\n", "Failed to clear running config item "
+				  "\"lxc.net.1.type\"");
+		goto on_error_destroy;
```
Comment (hallyn) [inline src/tests/livepatch.c]: Can you add a test having the container reboot, and make sure it comes up right?

And maybe also, separately, stop+start it?

File: src/tests/livepatch.c (hunk 2 above)
Comment (brauner) [reply src/tests/livepatch.c]: Yeah, this should be `s/is not defined/is defined/`. :)

File: src/tests/livepatch.c (hunk 3 above)
Comment (brauner) [reply src/tests/livepatch.c]: Yeah, can do.

File: src/lxc/lxccontainer.c (hunk 1 above)
Comment (brauner) [reply src/lxc/lxccontainer.c]: Yeah, I mean we can reverse the naming for the two functions and export the `do_fn()` method in `lxc.h`. Wouldn't matter much I'd say.

**Comment by brauner** [comment]
@hallyn, updated. Tests for reboot and stop + start added.

File: src/lxc/lxccontainer.c (hunk 1 above)
Comment (hallyn) [reply src/lxc/lxccontainer.c]: Yeah, let's leave it.  Thanks.

**Pull request merged**